metals/src/main/scala/scala/meta/internal/metals/onTypeRangeFormatters/OnTypeRangeFormatter.scala

```diff
@@ -0,0 +1,36 @@
+package scala.meta.internal.metals.onTypeRangeFormatters
+
+import scala.meta.tokens.Tokens
+
+import org.eclipse.lsp4j.FormattingOptions
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.Range
+import org.eclipse.lsp4j.TextEdit
+
+trait OnTypeRangeFormatter {
+  type StartPosition = meta.Position
+  type EndPosition = meta.Position
+
+  def onRangeContribute(
+      sourceText: String,
+      range: Range,
+      splitLines: Array[String],
```

Comment:
Do we need to pass `splitLines`? That's an implementation detail of a particular OnTypeRangeFormatter

Reply:
Since `splitLines` are used both in the `MultilineString` and in the `IndentOnPaste` formatters, I thought it would be more "efficient" to avoid having to repeat the same splitting operation of the entire sourceCode for both formatters.

Reply:
ok sure, might be needed then. Alternatively we can make it a lazy val in a case class that would contain more of the parameters, this way it will only be used if needed.

Reply:
Which parameters do you think are the best to keep in the formatting API, + the new case class?

Reply:
If we need everything then it can probably be all the existing parameters.

Reply:
I tried to do implement some sort of abstraction, it may need another iteration or two to be good enough, tho.